torch/csrc/jit/autodiff.cpp

```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+    JIT_ASSERT(output->stage() == stage);
+    graph_clone->registerOutput(val_map.at(output));
+  }
+
+  return graph_clone;
+}
+
+static std::unordered_map<Value*, int64_t> computeValueTopoOrder(const std::shared_ptr<Graph>& graph) {
+  int64_t next_nr = 0;
+  std::size_t stage = 0;
+  std::unordered_map<Value*, int64_t> order;
+  auto inputs_it = graph->inputs().begin();
+  auto inputs_end = graph->inputs().end();
+  auto nodes_it = graph->nodes().begin();
+  auto nodes_end = graph->nodes().end();
+
+  while (inputs_it != inputs_end || nodes_it != nodes_end) {
+    for (; inputs_it != inputs_end; ++inputs_it) {
+      Value *input = *inputs_it;
+      if (input->stage() != stage) break;
+      order[input] = next_nr++;
+    }
+    for (; nodes_it != nodes_end; ++nodes_it) {
+      Node *node = *nodes_it;
+      if (node->stage() != stage) break;
+      for (Value * output : node->outputs())
+        order[output] = next_nr++;
+    }
+    // NOTE: outputs are either inputs or are produced by nodes,
+    // so we don't need to handle them in any way.
+    stage++;
+  }
+  JIT_ASSERT(stage == graph->stage() + 1);
+  return order;
+}
+
+LiftedReverse lambdaLiftReverse(std::shared_ptr<Graph>& graph) {
+  static const auto is_stage_0 = [](Value *v) { return v->stage() == 0; };
+  static const auto is_stage_1 = [](Value *v) { return v->stage() == 1; };
+  // NOTE: in the comments here first stage == stage 0
+  JIT_ASSERT(graph->stage() == 1);
+
+  // --------------------------------------------------------------------------
+  // 1. Find values of stage 0 that need to ba captured.
+  // --------------------------------------------------------------------------
+  // First, we need to find all values that are produced in the first stage,
+  // and used in the second one. They will need to be added as inputs of the reverse
+  // graph, and some of them may also need to be appended as outputs of the primal graph.
+  std::unordered_set<Value*> extra_reverse_inputs;
+  for (Node * node : *graph) {
+    if (node->stage() == 0) continue;
+    for (Value * input : node->inputs()) {
+      if (input->stage() != 0) continue;
+      extra_reverse_inputs.insert(input);
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // 2. Establish an ordering on inputs/outputs.
+  // --------------------------------------------------------------------------
+  // We will need to sort primal outputs and reverse inputs, so that they maintain
+  // an order of:
+  //   [primal outputs], [primal temporaries], [reverse inputs]
+  // where each section is sorted in the same way as in the input graph.
+  //
+  // This corresponds exactly to taking a topological ordering on values,
+  // but artificially boosting outputs of the first stage to precede all
+  // other values.
+  auto topo_order = computeValueTopoOrder(graph);
+  auto io_compare = [&](Value *a, Value *b) { return topo_order.at(a) < topo_order.at(b); };
```

Comment:
It's too confusing to encode the ordering of the outputs as a sort. I can't glance at this and know if it is correct. It would be easier if the temporaries were added to the primal as part of the process of splitting the graph rather than trying to construct one giant graph here.

Reply:
I disagree. There is a certain ordering on the outputs we expect to get for both primal outputs and reverse inputs, and this is a simple way of enforcing them *consistently*, in a way such that `splitOffGraph` can remain a non-specialized helper, that doesn't really care if it's splitting the primal or dual. Moving that logic there would require adding a bunch of special cases, and it would certainly be harder for me to verify correctness. Here, you have a guarantee that if only this order is constructed correctly, then the outputs will look exactly as we would like them to.